contracts/bridge/contracts/common/message/OutboundLane.sol

```diff
@@ -0,0 +1,332 @@
+        // Nonce of the oldest message that we haven't yet pruned. May point to not-yet-generated
+        // message if all sent messages are already pruned.
+        uint256 oldest_unpruned_nonce;
+        // Nonce of the latest message, received by bridged chain.
+        uint256 latest_received_nonce;
+        // Nonce of the latest message, generated by us.
+        uint256 latest_generated_nonce;
+    }
+
+    /* State */
+
+    OutboundLaneNonce public outboundLaneNonce;
+
+    uint256 public confirmationFee = 0.1 ether; // how to set confirmation_fee
+
+    // MessageKey => MessageData
+    mapping(uint256 => MessageData) public messages;
+
+    /**
+     * @notice Deploys the OutboundLane contract
+     * @param _lightClientBridge The contract address of on-chain light client
+     * @param _thisChainPosition The thisChainPosition of outbound lane
+     * @param _bridgedChainPosition The bridgedChainPosition of outbound lane
+     * @param _lanePosition The lanePosition of outbound lane
+     * @param _oldest_unpruned_nonce The oldest_unpruned_nonce of outbound lane
+     * @param _latest_received_nonce The latest_received_nonce of outbound lane
+     * @param _latest_generated_nonce The latest_generated_nonce of outbound lane
+     */
+    constructor(
+        address _lightClientBridge,
+        uint256 _thisChainPosition,
+        uint256 _bridgedChainPosition,
+        uint256 _lanePosition,
+        uint256 _oldest_unpruned_nonce,
+        uint256 _latest_received_nonce,
+        uint256 _latest_generated_nonce
+    ) public MessageVerifier(_lightClientBridge, _thisChainPosition, _bridgedChainPosition, _lanePosition) {
+        outboundLaneNonce = OutboundLaneNonce(_oldest_unpruned_nonce, _latest_received_nonce, _latest_generated_nonce);
+        _setupRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    }
+
+    // Send message over lane.
+    function send_message(address targetContract, bytes calldata encoded) external payable override nonReentrant returns (uint256) {
+        require(hasRole(OUTBOUND_ROLE, msg.sender), "Lane: NotAuthorized");
+        require(outboundLaneNonce.latest_generated_nonce - outboundLaneNonce.latest_received_nonce <= MAX_PENDING_MESSAGES, "Lane: TooManyPendingMessages");
+        require(outboundLaneNonce.latest_generated_nonce < uint128(-1), "Lane: Overflow");
+        uint256 nonce = outboundLaneNonce.latest_generated_nonce + 1;
+        uint256 fee = msg.value;
+        outboundLaneNonce.latest_generated_nonce = nonce;
+        MessagePayload memory messagePayload = MessagePayload({
+            sourceAccount: msg.sender,
+            targetContract: targetContract,
+            encoded: encoded
+        });
+        uint256 key = encodeMessageKey(nonce);
+        // finally, save messageData in outbound storage and emit `MessageAccepted` event
+        messages[key] = MessageData({
+            payload: messagePayload,
+            fee: fee  // a lowest fee may be required and how to set it
+        });
```

Comment:
can we save nonce as this key directly? there are more gas if we calculate this key each time to read or write the message.